parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java

```diff
@@ -165,20 +165,41 @@ public <T> Boolean isNaN(BoundReference<T> ref) {
       }
 
       Set<T> dictionary = dict(id, comparatorForNaNPredicate(ref));
-      return dictionary.stream().anyMatch(NaNUtil::isNaN) ? ROWS_MIGHT_MATCH : ROWS_CANNOT_MATCH;
+      return dictionary.contains(naNValueForType(ref.type()))
+          ? ROWS_MIGHT_MATCH
+          : ROWS_CANNOT_MATCH;
     }
 
     @Override
     public <T> Boolean notNaN(BoundReference<T> ref) {
       int id = ref.fieldId();
 
+      if (mayContainNulls.get(id)) {
+        return ROWS_MIGHT_MATCH;
+      }
+
       Boolean hasNonDictPage = isFallback.get(id);
       if (hasNonDictPage == null || hasNonDictPage) {
         return ROWS_MIGHT_MATCH;
       }
 
       Set<T> dictionary = dict(id, comparatorForNaNPredicate(ref));
-      return dictionary.stream().allMatch(NaNUtil::isNaN) ? ROWS_CANNOT_MATCH : ROWS_MIGHT_MATCH;
+
+      if (dictionary.size() > 1 || !dictionary.contains(naNValueForType(ref.type()))) {
```

Comment:
I don't think this is correct. Parquet may have more than one NaN value because NaN values exist. I don't think Parquet guarantees that values are normalized to a single NaN. I think this should roll back both changes.

Reply:
I don't know how Parquet handles NaN values equality in dictionary, the reason I did this change is because  `dictionary` set is [using](https://github.com/apache/iceberg/blob/be0c8ddfd2a6d3c0f2e38b0b980d8fe73980ca3f/parquet/src/main/java/org/apache/iceberg/parquet/ParquetDictionaryRowGroupFilter.java#L184) `NaturalOrderComparator` which treats NaN values of the same type as equal, so I think there can only be at most one NaN value in this `dictionary`.